NEASQC/WP6_QNLP
# New dim reduction

## Scope
This PR modifies the dimensionality reduction module to add a functionality that allows the user to fit a dimensionality reduction object on a set of data X and then transform a set of data Y with this fitted object. Previously, this module only possessed a fit_transform functionality, that would fit on data X then transform data X, which is insufficient for our updated experiment pipeline for our alpha 3 and 4 models. 

Given that sklearn dim reduction objects work with numpy arrays, it made sense to change our embedder module to save embeddings as numpy arrays to a dataframe, and save the resulting vectorised dataset to a pickle file, which preserves python data types contrary to TSV, which we were using before. This PR makes the necessary modifications to the embedder module to ensure seamless interaction with the dimensionality reduction module.


## Implementation
New method implemented on dim_reduction.py, called 'fit'. This takes a dataframe\list of dataframes with vector embeddings and fits the dimensionality reduction object on the embeddings contained in those.

The method reduce_dimension has been modified to perform a transform operation, instead of fit_transform which it was performing before. This means that we can now fit on train+val, and transform train, val and test separately. 

The save_dataset methods in the embedder and dim_reduction modules have been modified to save dataframes as pickle files.

A number of methods in embedder and dim_reduction have received minor changes to ensure all embedding vectors are stored and read as numpy arrays.

## How to Test
Run the test_dim_reduction.py and test_embedder.py test modules. These have been modified to ensure the new changes are working properly.

## Emoji Guide
**For reviewers: Emojis can be added to comments to call out blocking versus non-blocking feedback.**
E.g: Praise, minor suggestions, or clarifying questions that don’t block merging the PR.
> 🟢 Nice refactor!
> 🟡 Why was the default value removed?
E.g: Blocking feedback must be addressed before merging.
> 🔴 This change will break something important

| | | |
| --- | --- | --- |
| Blocking | 🔴 ❌ 🚨 | Red |
| Non-blocking | 🟡 💡 🤔 💭 | Yellow, thinking, etc |
| Praise | 🟢 💚 😍 👍 🙌 | Green, hearts, positive emojis, etc |
